enzyme/Enzyme/InstructionBatcher.h

```diff
@@ -0,0 +1,242 @@
+      ValueToValueMapTy vmap;
+      Instruction *new_phi = placeholder->clone();
+      vmap[placeholder] = new_phi;
+
+      for (unsigned j = 0; j < phi.getNumIncomingValues(); ++j) {
+        Value *orig_block = phi.getIncomingBlock(j);
+        BasicBlock *new_block = cast<BasicBlock>(originalToNewFn[orig_block]);
+        Value *orig_val = phi.getIncomingValue(j);
+        Value *new_val;
+        if (isa<Constant>(orig_val)) {
+          new_val = orig_val;
+        } else {
+          new_val = vectorizedValues[orig_val][i];
+        }
+        vmap[placeholder->getIncomingValue(j)] = new_val;
+        vmap[new_block] = new_block;
+      }
+
+      RemapInstruction(new_phi, vmap, RF_NoModuleLevelChanges);
+      Instruction *placeholder = cast<Instruction>(vectorizedValues[&phi][i]);
+      ReplaceInstWithInst(placeholder, new_phi);
+      new_phi->setName(phi.getName());
+      vectorizedValues[&phi][i] = new_phi;
+    }
+  }
+
+  void visitSwitchInst(llvm::SwitchInst &inst) {
+    EmitFailure("SwitchConditionCannotBeVectorized", inst.getDebugLoc(), &inst,
+                "switch conditions have to be scalar values", inst);
+    llvm_unreachable("vectorized control flow is not allowed");
+  }
+
+  void visitBranchInst(llvm::BranchInst &branch) {
+    EmitFailure("BranchConditionCannotBeVectorized", branch.getDebugLoc(),
+                &branch, "branch conditions have to be scalar values", branch);
+    llvm_unreachable("vectorized control flow is not allowed");
+  }
+
+  void visitReturnInst(llvm::ReturnInst &ret) {
+    auto found = originalToNewFn.find(ret.getParent());
+    assert(found != originalToNewFn.end());
+    BasicBlock *nBB = dyn_cast<BasicBlock>(&*found->second);
+    IRBuilder<> Builder2 = IRBuilder<>(nBB);
+    Builder2.SetCurrentDebugLocation(DebugLoc());
+    ReturnInst *placeholder = cast<ReturnInst>(nBB->getTerminator());
+    SmallVector<Value *, 4> rets;
+
+    for (unsigned j = 0; j < ret.getNumOperands(); ++j) {
+      Value *op = ret.getOperand(j);
+      for (unsigned i = 0; i < width; ++i) {
+        Value *new_op;
+        if (toVectorize.count(op) != 0) {
+          new_op = vectorizedValues[op][i];
+        } else {
+          new_op = originalToNewFn[op];
+        }
+        rets.push_back(new_op);
+      }
+    }
+
+    if (ret.getNumOperands() != 0) {
+      auto ret = Builder2.CreateAggregateRet(rets.data(), width);
+      ret->setDebugLoc(placeholder->getDebugLoc());
+      placeholder->eraseFromParent();
+    }
+  }
+
+  void visitCallInst(llvm::CallInst &call) {
+    auto found = vectorizedValues.find(&call);
+    assert(found != vectorizedValues.end());
+    auto placeholders = found->second;
+    CallInst *placeholder = cast<CallInst>(placeholders[0]);
+    IRBuilder<> Builder2(placeholder);
+    Builder2.SetCurrentDebugLocation(DebugLoc());
+    Function *orig_func = getFunctionFromCall(&call);
+
+    auto isDefined = !orig_func->isDeclaration();
+    if (isDefined) {
```

Comment:
Style: I'd restructure this slightly to:
if (!isDefined) {
  visitInstruction();
  return;
}

...